modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1403 @@
+  }
+  return optval;
+}
+
+/*
+  Returns the value of the given socket option (see `getsockopt </https://pubs.opengroup.org/onlinepubs/9699919799/functions/getsockopt.html>`_)
+  on provided :type:`tcpConn`. The needed symbolic constants (SO_* etc.)
+  are defined in :mod:`Sys` module.
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+  :arg level: protocol level
+  :type level: `int(32)`
+  :arg optname: option to set.
+  :type optname: `int(32)`
+  :return: value of socket option
+  :rtype: `int(32)`
+  :throws SystemError: Upon incompatible arguments
+                        and socket.
+*/
+proc getSockOpt(ref socket: tcpConn, level: c_int, optname: c_int): int(32) throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname) ;
+}
+
+/*
+  See above
+*/
+proc getSockOpt(ref socket: tcpListener, level: c_int, optname: c_int): int(32) throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname) ;
+}
+
+/*
+  See above
+*/
+proc getSockOpt(ref socket: udpSocket, level: c_int, optname: c_int): int(32) throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname) ;
+}
+
+pragma "no doc"
+proc getSockOpt(socketFd:fd_t, level: c_int, optname: c_int, buflen: uint(16)) throws {
+  if buflen < 0 || buflen > 1024 {
+    throw new Error("getSockOpt buflen out of range");
+  }
+  else {
+    var len:socklen_t = buflen;
+    var buffer = c_calloc(c_uchar, buflen);
+    var err_out = sys_getsockopt(socketFd, level, optname, buffer:c_void_ptr, len);
+    if err_out != 0 {
+      c_free(buffer);
+      throw SystemError.fromSyserr(err_out, "Failed to get socket option");
+    }
+    return createBytesWithOwnedBuffer(buffer, len, buflen);
+  }
+}
+
+/*
+  Returns the value of the given socket option which is expected to be of type
+  :mod:`bytes <Bytes>` on provided :type:`tcpConn`. The needed symbolic constants (SO_* etc.)
+  are defined in :mod:`Sys` module.
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+  :arg level: protocol level
+  :type level: `int(32)`
+  :arg optname: option to set.
+  :type optname: `int(32)`
+  :return: value of socket option
+  :rtype: :mod:`bytes <Bytes>`
+  :throws SystemError: Upon incompatible arguments
+                        and socket.
+*/
+proc getSockOpt(ref socket: tcpConn, level: c_int, optname: c_int, buflen: uint(16)): bytes throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname, buflen) ;
+}
+
+/*
+  See Above
+*/
+proc getSockOpt(ref socket: tcpListener, level: c_int, optname: c_int, buflen: uint(16)): bytes throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname, buflen) ;
+}
+
+/*
+  See Above
+*/
+proc getSockOpt(ref socket: udpSocket, level: c_int, optname: c_int, buflen): bytes throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname, buflen) ;
+}
```

Comment:
If we make any changes to `setSockOpt`, we should make symmetric changes to `getSockOpt`